### DISCUSSION TITLE
Support Client-Only routes.

### DISCUSSION THREAD

**Discussion by Trumspringa** [withastro/roadmap#947] [Proposal]
# Summary

Astro requires SSR for parameterized routes to be resolved dynamically. This should be possible without the need of SSR with client-only routes. As seen in GatsbyJS and NextJS

# Background & Motivation

I want to have a 100% static site with a few routes that are behind authentication and communicate with a Supabase database. This is currently possible in Gatsby using ["client-only routes"](https://www.gatsbyjs.com/docs/how-to/routing/client-only-routes-and-user-authentication/) and I also believe it's possible using the Pages router in Next and coming soon to their App router.

It would remove the need for adapters in this use case where SSR is otherwise not wanted.

# Goals

To be able to have dynamic routes controlled by an SPA router that takes over any routes. For example a page at /app/[...].astro could just render an <App /> component with react-router, solid-router etc and the client-only directive. The [...] syntax is what gatsby uses. Hosting would also need to configure any redirects from /app/* to /app/[...]/index.html. But it seems like it should be doable with Astro.

Are there any roadblocks in implementing this?

**Comment by ematipico** [comment]
Gatsby is a React framework, so you'd have client side logic too, out of the box.

Astro doesn't impose any client decisions. So if you want to have a mini spa, you can already do this today by using a dynamic route, and render your client side framework there.

**Reply by Trumspringa** [reply]
Correct. But the dynamic route needs to be SSR. I'm suggesting to make it possible to do without SSR like Gatsby and NextJS allow.

**Reply by ematipico** [reply]
Why do you need an on-demand page? I use a simple static page an it works like a charm

**Reply by Trumspringa** [reply]
Because mine uses dynamic routes. Do you use getStaticPaths in yours?

I basically have an app directory where I have one astro page [..all].astro, I tried just having a <App /> component there which is the solid-js router with multiple routes, thinking that anything /app/* would be handled by this page. But got an error saying that a getStaticPaths function needs to be exported for dynamic routes, meaning you need to know the paths at build time. 

I have a route for example /app/todos/:id where id is dynamic and I want it to be resolved client side by the router not by using getStaticPaths.

[The Astro documentation says your project needs to be either SSR or Hybrid with the dynamic routes page being SSR in order to achieve this.](https://docs.astro.build/en/guides/routing/#_top)

"In Astro’s default static output mode, these pages are generated at build time, and so you must predetermine the list of authors that get a corresponding file. In SSR mode, a page will be generated on request for any route that matches".

In order for my example to work I would have to read the database and get all todoIds then use getStaticPaths to generate the paths that will be handled by the router. I assume that would work although I haven't tried. But I would like the page to just be static and render the <App /> component for all routes in that directory. Gatsby and NextJS can both do it.

I'm interested in why that needs SSR in Astro and whether it can be achieved in static mode. As that would negate the need for a server.

**Comment by Trumspringa** [comment]
The docs state that it is possible ["Static sites can include Astro islands for interactive UI components **or even entire embedded client-side rendered apps!** written in the UI framework of your choice in an otherwise static, pre-rendered page".](https://docs.astro.build/en/basics/rendering-modes/#pre-rendered) But doesn't provide any information on how to achieve it.

Not sure how it can be done when dynamic routes in static mode must export the getStaticPaths function and all paths need to be known at build time.

Here's my code that currently only works if the page is **not** pre-rendered. This is in `pages/app/[...path].astro`
```javascript
---
export const prerender = false
import BaseLayout from "../../components/BaseLayout.astro";
import App from '../../components/App.jsx'
---

<BaseLayout>
        <App client:only="solid-js" />
</BaseLayout>
```

And a simplified version of my App component is:

```javascript
import { Router, Route } from "@solidjs/router";
import Layout from './Layout'
import Todos from './Todos'
import Todo from './Todo'
import Profile from './Profile'

export default function App() { 
    return (
        <Router root={Layout}>
            <Route path="/app/todos" component={Todos} />
            <Route path="/app/todos/:id" component={Todo} />
            <Route path="/app/profile" component={Profile} />
        </Router>
    )
}
```

**Comment by Trumspringa** [comment]
Finally found that it was an issue with my redirects on deployment because of trailing slashes.

Closing the discussion.

**Reply by mortbauer** [reply]
How did you get around the getStaticPaths function?

**Reply by Trumspringa** [reply]
You can do it just by enabling hybrid mode and have 'export const prerender = false' at the top of the file like above. This requires a server. But if like in my case you want a completely static site then you can do it by having a page just client render a component that has a router like I did above and instead of Astro serving the page to any matching route like it does in SSR you need to configure your hosting to rewrite urls so that the client side router can control what page to show. So you can remove 'export const prerender = false' and have your output in your config just be 'static'.  Then instead of using [...path].astro just use a normal page name, I use app.astro and finally for example Netlify has a _redirects file where you can configure rewrites.

The reason it didn't work for me initially was that I had /app/* /app 200 in my _redirects file but there's a trailing slash when the files are built so changing it to /app/* /app/ 200 fixed it. Giving the 200 status code means that it's a rewrite rather than a redirect, so the url won't change but the router page will still be served for any matching route.

**Reply by mortbauer** [reply]
Many thanks for the quick reply, I'm also going for static anyways and using nginx for the "rewrites", managed it to work.

**Reply by Knight-H** [reply]
since the rewrites work on the server, how would this work in the dev environment?

**Reply by Knight-H** [reply]
one hack i can think of is 
(1) re-enabling hybrid mode
(2) use the slug /app/[...path].astro with prerender=false and return Astro.rewrite("/app"), which will allow the App client component to be displayed 

And reverse everything back to static mode in deployment... which I think creating 2 projects of 1 SPA and 1 SSG and letting the server handle rewrites may be easier (unless there's an easier way to manage this).

**Reply by Trumspringa** [reply]
> since the rewrites work on the server, how would this work in the dev environment?

In my case netlify dev handles it based on my _redirects file. I think Cloudflare Pages also uses _redirects while Vercel uses a vercel.json file.
I imagine it should be straightforward to use NGINX for your dev server for rewrites too if you're not using Cloudflare, Netlify, Vercel etc. But I'm not familiar with using it.

> one hack i can think of is
(1) re-enabling hybrid mode
(2) use the slug /app/[...path].astro with prerender=false and return Astro.rewrite("/app"), which will allow the App client component to be displayed

You wouldn't need Astro.rewrite in this case as the dynamic path would always be handled by the [...path].astro file in SSR mode.

It would be very cumbersome to have to switch the rendering modes as I don't think you can change it based on import.meta.env.MODE being development or production and I'm also not sure if there's anyway to ignore a file being built, so after you change the rendering mode back to 'static' the build will fail as the /app/[...path].astro file doesn't export a getStaticPaths function.

**Reply by phuoccss** [reply]
For development alternatively you can set up a vite plugin.

```
   vite: {
    plugins: [
      {
        name: 'rewrite',
        configureServer(serve) {
          serve.middlewares.use((req, _, next) => {
            if (req.url?.startsWith('/app/')) {
              req.url = '/app'
            }
            next()
          })
        },
      },
    ],
  }
```

**Reply by erikdemarco** [reply]
> In my case netlify dev handles it based on my _redirects file.

What do you mean by "netlify dev"? Is it netlify adapter for astro? If I use cloudflare adapter will it gives the same result?

Where do you put the _redirect file? I'm public folder? And astro automatically handles the redirect in dev environment?

And why you close this issue? If accepted it will be way easier
